src/main/cpp/apple_fsnotifier.cpp

```diff
@@ -62,9 +62,15 @@ static void callback(ConstFSEventStreamRef streamRef,
 
     for (int i = 0; i < numEvents; i++) {
         FSEventStreamEventFlags flags = eventFlags[i];
-        log_fine(details->env, "~~~~ Event flags: 0x%x for %s\n", flags, paths[i]);
+        log_fine(details->env, "Event flags: 0x%x for %s", flags, paths[i]);
         jint type;
-        if (IS_SET(flags, kFSEventStreamEventFlagMustScanSubDirs)) {
+        if (IS_SET(flags, kFSEventStreamEventFlagHistoryDone)) {
+            continue;
+        } else if (IS_ANY_SET(flags,
+                kFSEventStreamEventFlagRootChanged
```

Comment:
Shall we also say if the flags are equal to `kFSEventStreamEventFlagNone` - which is `0x00000000`, then we return `INVALIDATE`? IIUC, that was the legacy event before the API was able to send better events.

Reply:
Hm, that's interesting. I wonder how useful that would be, though, as we'd effectively have two states: 1) nothing's changed, 2) everything's changed. This would mostly eliminate the utility of VFS retention, no? It is more or less the same outcome as we have now via `UNKNOWN`.

The more detailed API has been around since OS X Lion 10.7, which was released in 2011. I don't expect to encounter even earlier versions of macOS with VFS retention enabled. If users report a problem, we can always explicitly disable the functionality on those OS versions.

WDYT?

Reply:
That's fine, we can also leave it out.